tests/tb_device_mqtt_client_tests.py

```diff
@@ -82,19 +86,425 @@ def test_send_telemetry_and_attr(self):
+    def __init__(self, value):
+        self.value = value
+
+
+@unittest.skipUnless(has_rc(), "TBPublishInfo.rc() is missing from your local version of tb_device_mqtt.py")
+class TBPublishInfoTests(unittest.TestCase):
+    def test_rc_single_reasoncodes_zero(self):
+        message_info_mock = MagicMock()
+        message_info_mock.rc = FakeReasonCodes(0)
+
+        publish_info = TBPublishInfo(message_info_mock)
+        self.assertEqual(publish_info.rc(), 0)  # TB_ERR_SUCCESS
+
+    def test_rc_single_reasoncodes_nonzero(self):
+        message_info_mock = MagicMock()
+        message_info_mock.rc = FakeReasonCodes(128)
+
+        publish_info = TBPublishInfo(message_info_mock)
+        self.assertEqual(publish_info.rc(), 128)
+
+    def test_rc_single_int_nonzero(self):
+        message_info_mock = MagicMock()
+        message_info_mock.rc = 2
+
+        publish_info = TBPublishInfo(message_info_mock)
+        self.assertEqual(publish_info.rc(), 2)
+
+    def test_rc_list_all_zero(self):
+        mi1 = MagicMock()
+        mi1.rc = FakeReasonCodes(0)
+        mi2 = MagicMock()
+        mi2.rc = FakeReasonCodes(0)
+
+        publish_info = TBPublishInfo([mi1, mi2])
+        self.assertEqual(publish_info.rc(), 0)
+
+    def test_rc_list_mixed(self):
+        mi1 = MagicMock()
+        mi1.rc = FakeReasonCodes(0)
+        mi2 = MagicMock()
+        mi2.rc = FakeReasonCodes(128)
+
+        publish_info = TBPublishInfo([mi1, mi2])
+        self.assertEqual(publish_info.rc(), 128)
+
+    def test_rc_list_int_nonzero(self):
+        mi1 = MagicMock()
+        mi1.rc = 0
+        mi2 = MagicMock()
+        mi2.rc = 4
+
+        publish_info = TBPublishInfo([mi1, mi2])
+        self.assertEqual(publish_info.rc(), 4)
+
+    def test_mid_single(self):
+        message_info_mock = MagicMock()
+        message_info_mock.mid = 123
+
+        publish_info = TBPublishInfo(message_info_mock)
+        self.assertEqual(publish_info.mid(), 123)
+
+    def test_mid_list(self):
+        mi1 = MagicMock()
+        mi1.mid = 111
+        mi2 = MagicMock()
+        mi2.mid = 222
+
+        publish_info = TBPublishInfo([mi1, mi2])
+        self.assertEqual(publish_info.mid(), [111, 222])
+
+    @patch('logging.getLogger')
+    def test_get_single_no_exception(self, mock_logger):
+        message_info_mock = MagicMock()
+        publish_info = TBPublishInfo(message_info_mock)
+        publish_info.get()
+
+        message_info_mock.wait_for_publish.assert_called_once_with(timeout=1)
+        mock_logger.return_value.error.assert_not_called()
+
+    @patch('logging.getLogger')
+    def test_get_list_no_exception(self, mock_logger):
+        mi1 = MagicMock()
+        mi2 = MagicMock()
+        publish_info = TBPublishInfo([mi1, mi2])
+        publish_info.get()
+
+        mi1.wait_for_publish.assert_called_once_with(timeout=1)
+        mi2.wait_for_publish.assert_called_once_with(timeout=1)
+        mock_logger.return_value.error.assert_not_called()
```

Comment:
As it was mentioned before, it is a bad practice to base tests on log messages or calls to loggers/sleeps and other, not related to main functionality. Because these messages or calls may be changed without affecting the main logic, but test will fail.